src/braket/experimental/autoqasm/api.py

```diff
@@ -487,8 +541,139 @@ def _wrap_for_oqpy_gate(
+        qubits=qubits,
+        angles=angles,
+        oqpy_program=program_conversion_context.get_oqpy_program(),
+    )
+
+
+def _validate_calibration_args(
+    gate_function: Callable,
+    decorator_args: Dict[str, Union[Qubit, float]],
+    func_args: aq_program.GateArgs,
+) -> None:
+    """Validate the arguments passed to the calibration decorator and function.
+
+    Args:
+        gate_function (Callable): The gate function which calibration is being defined.
+        decorator_args (Dict[str, Union[Qubit, float]]): The calibration decorator arguments.
+        func_args (aq_program.GateArgs): The gate function arguments.
+    """
+    gate_args = _get_gate_args(gate_function)
+    gate_args_names = [var.name for var in gate_args._args]
+    func_args_names = [var.name for var in func_args._args]
+    decorator_args_names = decorator_args.keys()
+
+    # validate the name of args
+    if not set(gate_args_names) == set(decorator_args_names) | set(func_args_names):
+        raise errors.InvalidCalibrationDefinition(
+            "The union of calibration decorator arguments and function arguments must match the"
+            " gate arguments."
+        )
+
+    if any(name in decorator_args_names for name in func_args_names):
+        raise errors.InvalidCalibrationDefinition(
+            "The function arguments must not duplicate any argument in the calibration decorator."
+        )
+
+    # validate the type of args
+    for qubit_arg in gate_args.qubits:
+        if qubit_arg.name in decorator_args_names and not is_qubit_identifier_type(
+            decorator_args[qubit_arg.name]
+        ):
+            raise errors.ParameterTypeError(
+                f'Parameter "{qubit_arg.name}" must have a type of aq.Qubit.'
+            )
+        if qubit_arg.name in func_args_names and qubit_arg.name not in [
+            var.name for var in func_args.qubits
+        ]:
+            raise errors.ParameterTypeError(
+                f'Parameter "{qubit_arg.name}" must have a type hint of aq.Qubit.'
+            )
+
+    for angle_arg in gate_args.angles:
+        if angle_arg.name in decorator_args_names and not isinstance(
+            decorator_args[angle_arg.name], float
+        ):
+            raise errors.ParameterTypeError(
+                f'Parameter "{angle_arg.name}" must have a type of float.'
+            )
+        if angle_arg.name in func_args_names and angle_arg.name not in [
+            var.name for var in func_args.angles
+        ]:
+            raise errors.ParameterTypeError(
+                f'Parameter "{angle_arg.name}" must have a type hint of float.'
+            )
+
+
+def _categorize_calibration_decorator_args(decorator_kwargs) -> Tuple[List[str], List[str]]:
+    """Categorize the calibration decorator arguments into qubit and angle.
+
+    Args:
+        decorator_kwargs (Dict[str, Any]): The calibration decorator arguments.
+
+    Returns:
+        Tuple[List[str], List[str]]: The qubit and angle names.
```

Comment:
I wonder if it would be cleaner to return a `GateArgs` object from this function, instead of two lists of strings?